pkg/clustersql/settings_test.go

```diff
@@ -0,0 +1,245 @@
+		require.Equal(t, ErrInvalidClusterSettingName, errors.Cause(err))
+	})
+
+	t.Run("returns error when exec fails", func(t *testing.T) {
+		mock.
+			ExpectExec("SET CLUSTER SETTING bogus_setting = \\$1").
+			WithArgs("nope").
+			WillReturnError(errors.New("boom"))
+
+		err := SetClusterSetting(context.Background(), db, "bogus_setting", "nope")
+		require.EqualError(t, errors.Cause(err), "boom")
+	})
+}
+
+func TestRangeMoveDuration(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	require.NoError(t, err)
+	defer db.Close()
+
+	ctx := context.Background()
+
+	stubSettings := func(settings map[string]string) {
+		// map keys are not ordered
+		mock.MatchExpectationsInOrder(false)
+
+		for k, v := range settings {
+			mock.
+				ExpectQuery("SHOW CLUSTER SETTING " + k).
+				WillReturnRows(sqlmock.NewRows([]string{"name"}).AddRow(v))
+		}
+	}
+
+	t.Run("calculates the time it takes for a range to move nodes", func(t *testing.T) {
+		tests := []struct {
+			name             string
+			maxRebalanceRate string
+			maxRecoveryRate  string
+			zones            []Zone
+			exp              time.Duration
+		}{
+			{
+				name:             "rebalance rate faster than recovery",
+				maxRebalanceRate: "10MB",
+				maxRecoveryRate:  "6MB",
+				zones: []Zone{
+					{Target: "zone-1", Config: ZoneConfig{RangeMaxBytes: 10}},
+					{Target: "zone-2", Config: ZoneConfig{RangeMaxBytes: 20}},
+				},
+				exp: time.Duration(10/1000000) * time.Second,
+			},
+			{
+				name:             "recovery rate faster than rebalance",
+				maxRebalanceRate: "1MB",
+				maxRecoveryRate:  "2MB",
+				zones: []Zone{
+					{Target: "zone-1", Config: ZoneConfig{RangeMaxBytes: 10}},
+					{Target: "zone-2", Config: ZoneConfig{RangeMaxBytes: 20}},
+				},
+				exp: time.Duration(20/1000000) * time.Second,
+			},
+			{
+				name:             "min range size is used",
+				maxRebalanceRate: "1MB",
+				maxRecoveryRate:  "2MB",
+				zones: []Zone{
+					{Target: "zone-1", Config: ZoneConfig{RangeMaxBytes: 20}},
+					{Target: "zone-2", Config: ZoneConfig{RangeMaxBytes: 30}},
+				},
+				exp: time.Duration(20/2000000) * time.Second,
+			},
+		}
+
+		for _, tt := range tests {
+			stubSettings(map[string]string{
+				"kv.snapshot_rebalance.max_rate": tt.maxRebalanceRate,
+				"kv.snapshot_recovery.max_rate":  tt.maxRecoveryRate,
+			})
+
+			d, err := RangeMoveDuration(ctx, db, tt.zones...)
+			require.Equal(t, tt.exp, d)
+			require.NoError(t, err)
+		}
+	})
+
+	t.Run("returns error when MaxRangeSize is 0 for all zones", func(t *testing.T) {
+		stubSettings(map[string]string{
+			"kv.snapshot_rebalance.max_rate": "10MB",
+			"kv.snapshot_recovery.max_rate":  "10MB",
+		})
+
+		d, err := RangeMoveDuration(ctx, db, Zone{
+			Target: "us-east1-c",
+			Config: ZoneConfig{RangeMaxBytes: 0},
+		})
+
+		require.Zero(t, d)
+		require.EqualError(t, err, "no maximum range size found")
+	})
+
+	t.Run("return error when fetching ZoneConfigs fails", func(t *testing.T) {
```

Comment:
I don't understand enough of this, but are we testing everything together?   If we are testing this much functionality I am wondering if we need to refactor the code in smaller chunks.  Not in this PR, but in another.

Reply:
I tend to agree here. Currently this package has no unit tests, so I've been working on the idea of getting tests in place, seeing what's ugly/undesirable, and then refactoring from there.